### PR TITLE
Add CISPO loss type support for LigerFusedLinearGRPOLoss

### PR DESCRIPTION
**Pull request by yukiu00**
## Summary
Resolve: https://github.com/linkedin/Liger-Kernel/issues/1057

* Add **CISPO** (`loss_type="cispo"`) support to **`LigerFusedLinearGRPOLoss`** (chunked loss path)
* Enable TRL's **`GRPOTrainer`** to work with `use_liger_kernel=True` and `loss_type="cispo"`

### Background / Motivation

CISPO (Clipped Importance Sampling Policy Optimization) is a loss variant proposed in the **MiniMax-M1** technical report. It clips the importance sampling ratio with **only an upper bound** and **detaches it from gradient computation**.

TRL added `loss_type="cispo"` to `GRPOTrainer`, but Liger Kernel did not support it, causing errors when using `use_liger_kernel=True` with `loss_type="cispo"`.

### Changes

**`src/liger_kernel/chunked_loss/grpo_loss.py`**

* Add CISPO loss matching TRL's implementation
  * Clip importance sampling ratio with **upper bound only** and **detach**:
    ```python
    clamped_ratios = torch.clamp(coef_1, max=epsilon_high).detach()
    ```
* Use **DAPO-style normalization** for CISPO reduction (consistent with TRL)
* Add CISPO-specific clip metric for logging compatibility:
  * Count tokens where `(coef_1 > epsilon_high) & (advantages > 0)`

**`src/liger_kernel/transformers/grpo_loss.py`**

* Add CISPO reduction logic (uses same normalizer as DAPO)
* Raise explicit error for Triton GRPO loss path (CISPO not supported there)

**`ops/grpo_loss` (Triton fused path)**

* CISPO is **not implemented** in `ops/grpo_loss` in this PR
* `loss_type="cispo"` is **only supported via chunked loss path** (Triton fused support is a follow-up)

**`test/chunked_loss/test_grpo_loss.py`**

* Add CISPO to torch reference implementation (`TorchLMHeadGRPO`)
* Add `"cispo"` to parameterized test cases to verify parity with reference

### References

* MiniMax-M1 (CISPO introduction): https://arxiv.org/abs/2506.13585
* DAPO (normalization / reduction reference): https://arxiv.org/abs/2503.14476
* TRL CISPO implementation: https://github.com/huggingface/trl/blob/035c3ff151b953ca72cdfe0ee966bc1469a26fde/trl/trainer/grpo_trainer.py#L2030

## Testing Done

- Hardware Type: RTX3090 24GB (NVIDIA Ampere)
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: test/chunked_loss/test_grpo_loss.py [hunk 1]
```diff
@@ -77,28 +78,44 @@ def compute_per_token_components(
...
-        else:  # sequence level
-            # For sequence level, coef_1 is shape (B, 1), advantages is shape (B,)
-            seq_advantages = advantages
-            is_clipped = ((coef_1.squeeze(-1) < 1 - epsilon_low) & (seq_advantages < 0)) | (
-                (coef_1.squeeze(-1) > 1 + epsilon_high) & (seq_advantages > 0)
-            )
-            is_clipped = is_clipped.unsqueeze(1).expand_as(attention_mask)
+        if loss_type == "cispo":
+            # CISPO: clipped when coef_1 > epsilon_high AND advantages > 0
+            # Reference: https://github.com/huggingface/trl/blob/035c3ff151b953ca72cdfe0ee966bc1469a26fde/trl/trainer/grpo_trainer.py#L2073
```
Comment (Mecoli1219) [inline test/chunked_loss/test_grpo_loss.py]: Similar to the previous comment. The reference seems to be incorrect. Might be this one: [http://github.com/huggingface/trl/blob/035c3ff151b953ca72cdfe0ee966bc1469a26fde/trl/trainer/grpo_trainer.py#L2120-L2124](https://github.com/huggingface/trl/blob/035c3ff151b953ca72cdfe0ee966bc1469a26fde/trl/trainer/grpo_trainer.py#L2120-L2124)

File: src/liger_kernel/chunked_loss/grpo_loss.py [hunk 2]
```diff
@@ -106,16 +112,26 @@ def ppo_loss_fn(
...
-            )
-        else:  # sequence level
-            # For sequence level, coef_1 is shape (B, 1), advantages is shape (B,)
-            is_clipped = ((coef_1.squeeze(-1) < 1 - epsilon_low) & (advantages < 0)) | (
-                (coef_1.squeeze(-1) > 1 + epsilon_high) & (advantages > 0)
-            )
-            is_clipped = is_clipped.unsqueeze(1).expand_as(attention_mask)
+        if loss_type == "cispo":
+            # CISPO: clipped when coef_1 > epsilon_high AND advantages > 0
+            # Reference: https://github.com/huggingface/trl/blob/035c3ff151b953ca72cdfe0ee966bc1469a26fde/trl/trainer/grpo_trainer.py#L2073
```
Comment (Mecoli1219) [inline src/liger_kernel/chunked_loss/grpo_loss.py]: Do you mean this: http://github.com/huggingface/trl/blob/035c3ff151b953ca72cdfe0ee966bc1469a26fde/trl/trainer/grpo_trainer.py#L2120-L2124

File: src/liger_kernel/transformers/grpo_loss.py [hunk 3]
```diff
@@ -28,6 +28,8 @@ def triton_grpo_loss(
         raise ValueError(
             f"Triton GRPO loss only supports token-level importance sampling. Got {importance_sampling_level}."
         )
+    if loss_type == "cispo":
+        raise ValueError("Triton GRPO loss does not support loss_type='cispo'. Use the chunked GRPO loss path.")
```
Comment (Mecoli1219) [inline src/liger_kernel/transformers/grpo_loss.py:32]: Do we need to throw error here given that the `_reduce_grpo_loss` is updated to support `cispo`?

File: src/liger_kernel/chunked_loss/grpo_loss.py (hunk 2 above)
Comment (yukiu00) [reply src/liger_kernel/chunked_loss/grpo_loss.py]: Thanks for catching this! Fixed in https://github.com/linkedin/Liger-Kernel/pull/1054/changes/2a9517f42d0e1bc528fabed3f216b23796dc6c37.

File: test/chunked_loss/test_grpo_loss.py (hunk 1 above)
Comment (yukiu00) [reply test/chunked_loss/test_grpo_loss.py]: Thanks for catching this! Fixed in https://github.com/linkedin/Liger-Kernel/pull/1054/changes/2a9517f42d0e1bc528fabed3f216b23796dc6c37.

File: src/liger_kernel/transformers/grpo_loss.py (hunk 3 above)
Comment (yukiu00) [reply src/liger_kernel/transformers/grpo_loss.py:32]: The `_reduce_grpo_loss` function only handles loss aggregation/reduction, but the actual per-token loss computation happens in the Triton kernel (`GrpoLossFunction`).

CISPO uses a fundamentally different formula:
- CISPO: `-clamp(ratio, max=eps_high).detach() * advantages * per_token_logps`
- Standard GRPO: `-min(ratio * adv, clamp(ratio, 1-eps, 1+eps) * adv)`

Since the Triton kernel doesn't implement CISPO's per-token loss calculation, the ValueError is needed to direct users to the chunked loss path (`LigerFusedLinearGRPOLoss`), which does support CISPO.

Triton fused path support for CISPO would be a follow-up.

Comment (Mecoli1219) [reply src/liger_kernel/transformers/grpo_loss.py:32]: I see. That makes sense. If that is the case, I think we should move the change for `_reduce_grpo_loss` to the future PR that support triton-fused CISPO since `_reduce_grpo_loss` is only used in `triton_grpo_loss`. 
@Tcc0403, wdyt?

Comment (yukiu00) [reply src/liger_kernel/transformers/grpo_loss.py:32]: @Mecoli1219 

Thanks for the suggestion. I've removed the CISPO case from _reduce_grpo_loss since it's currently unreachable (blocked by the ValueError). This will be added back when we implement Triton fused CISPO support in a follow-up PR.
Fixed in b521461.

**Review by Mecoli1219 — APPROVED** [review]
Thanks @yukiu00. The PR looks good to me. Let's wait @Tcc0403 to run the test before merging it.

File: src/liger_kernel/chunked_loss/grpo_loss.py
Comment (Tcc0403) [inline src/liger_kernel/chunked_loss/grpo_loss.py]:
```py
def clip_coef_fn(coef, epsilon_low, epsilon_high, loss_type):
    if loss_type == "cispo":
        # CISPO: clip and detach the importance weights
        upper_bound = epsilon_high
        lower_bound = None
        clipped_coef = torch.clamp(coef, lower_bound, upper_bound).detach()
        is_lower_clipped = False
        is_upper_clipped = coef > upper_bound
    else:
        upper_bound = 1 + epsilon_high
        lower_bound = 1 - epsilon_low
        clipped_coef = torch.clamp(coef, lower_bound, upper_bound)
        is_lower_clipped = coef < lower_bound
        is_upper_clipped = coef > upper_bound
    return clipped_coef, is_lower_clipped, is_upper_clipped
```

File: src/liger_kernel/chunked_loss/grpo_loss.py [hunk 4]
```diff
@@ -67,10 +67,16 @@ def ppo_loss_fn(
...
+        if loss_type == "cispo":
+            # CISPO: clip and detach the importance weights, multiply by log probs
+            # Reference: https://github.com/huggingface/trl/blob/035c3ff151b953ca72cdfe0ee966bc1469a26fde/trl/trainer/grpo_trainer.py#L2030
+            clamped_ratios = torch.clamp(coef_1, max=epsilon_high).detach()
+            per_token_loss = -clamped_ratios * advantages.unsqueeze(1) * per_token_logps
+        else:
+            coef_2 = clip_coef_fn(coef_1, epsilon_low, epsilon_high)
+            per_token_loss1 = coef_1 * advantages.unsqueeze(1)
+            per_token_loss2 = coef_2 * advantages.unsqueeze(1)
+            per_token_loss = -torch.min(per_token_loss1, per_token_loss2)
```
Comment (Tcc0403) [inline src/liger_kernel/chunked_loss/grpo_loss.py:91]:
```suggestion
        coef_2, is_lower_clipped, is_upper_clipped = clip_coef_fn(coef_1, epsilon_low, epsilon_high, loss_type)
        if loss_type == "cispo":
            # CISPO: clip and detach the importance weights, multiply by log probs
            # Reference: https://github.com/huggingface/trl/blob/035c3ff151b953ca72cdfe0ee966bc1469a26fde/trl/trainer/grpo_trainer.py#L2030
            per_token_loss = -clamped_ratios * advantages.unsqueeze(1) * per_token_logps
        else:
            per_token_loss1 = coef_1 * advantages.unsqueeze(1)
            per_token_loss2 = coef_2 * advantages.unsqueeze(1)
            per_token_loss = -torch.min(per_token_loss1, per_token_loss2)
```

File: src/liger_kernel/chunked_loss/grpo_loss.py [hunk 5]
```diff
@@ -106,16 +112,26 @@ def ppo_loss_fn(
...
+            if importance_sampling_level == "token":
+                is_clipped = ((coef_1 < 1 - epsilon_low) & (advantages.unsqueeze(1) < 0)) | (
+                    (coef_1 > 1 + epsilon_high) & (advantages.unsqueeze(1) > 0)
+                )
+            else:  # sequence level
+                # For sequence level, coef_1 is shape (B, 1), advantages is shape (B,)
+                is_clipped = ((coef_1.squeeze(-1) < 1 - epsilon_low) & (advantages < 0)) | (
+                    (coef_1.squeeze(-1) > 1 + epsilon_high) & (advantages > 0)
+                )
+                is_clipped = is_clipped.unsqueeze(1).expand_as(attention_mask)
```
Comment (Tcc0403) [inline src/liger_kernel/chunked_loss/grpo_loss.py]:
```suggestion
        # Adjust clipping metric calculation based on importance sampling level
        if importance_sampling_level == "token":
            is_clipped = (is_lower_clipped & (advantages.unsqueeze(1) < 0)) | (
                is_upper_clipped & (advantages.unsqueeze(1) > 0)
            )
        else:  # sequence level
            # For sequence level, coef_1 is shape (B, 1), advantages is shape (B,)
            is_clipped = (is_lower_clipped & (advantages < 0)) | (is_upper_clipped & (advantages > 0))
            is_clipped = is_clipped.expand_as(attention_mask)
```

**Review by Tcc0403 — COMMENTED** [review]
Thank you so much! Current changes passed all tests, just a little bit refactor for readability

**Comment by yukiu00** [comment]
@Tcc0403 

Thanks for the review! I've refactored clip_coef_fn as suggested to return (clipped_coef, is_lower_clipped, is_upper_clipped) and unified the clipping logic for both CISPO and other loss types.

fca10e0

**Comment by Tcc0403** [comment]
There are some errors where `importance_sampling_level == true`. Do you mind if I edit your PR directly?

File: test/chunked_loss/test_grpo_loss.py [hunk 6]
```diff
@@ -77,28 +78,43 @@ def compute_per_token_components(
...
+            is_clipped = (is_lower_clipped & (expanded_advantages < 0)) | (is_upper_clipped & (expanded_advantages > 0))
         else:  # sequence level
             # For sequence level, coef_1 is shape (B, 1), advantages is shape (B,)
-            seq_advantages = advantages
-            is_clipped = ((coef_1.squeeze(-1) < 1 - epsilon_low) & (seq_advantages < 0)) | (
-                (coef_1.squeeze(-1) > 1 + epsilon_high) & (seq_advantages > 0)
-            )
-            is_clipped = is_clipped.unsqueeze(1).expand_as(attention_mask)
+            is_clipped = (is_lower_clipped & (advantages < 0)) | (is_upper_clipped & (advantages > 0))
+            is_clipped = is_clipped.expand_as(attention_mask)
```
Comment (Tcc0403) [inline test/chunked_loss/test_grpo_loss.py]:
```suggestion
            is_clipped = (is_lower_clipped & (advantages.unsqueeze(1) < 0)) | (is_upper_clipped & (advantages.unsqueeze(1) > 0))
            is_clipped = is_clipped.expand_as(attention_mask)
```
We need to unsequeeze advenatages so that the first term `(is_lower_clipped & (advantages.unsqueeze(1) < 0))` can be expanded correctly since `is_lower_clipped` has no dimension.

Before fix, shapes would be:
```
(is_lower_clipped & (advantages < 0)) | (is_upper_clipped & (advantages > 0))
= (((None) & (B,)) | ((B, 1) & (B,)) 
= (B,) | (B, 1)
= (B, B)
```

After fix:
```
(is_lower_clipped & (advantages.unsqueeze(1) < 0)) | (is_upper_clipped & (advantages.unsqueeze(1) > 0))
= ((None & (B, 1) | ((B,1) & (B,))
= (B, 1) | (B, 1)
= (B, 1)
```

File: test/chunked_loss/test_grpo_loss.py [hunk 7]
```diff
@@ -77,28 +78,43 @@ def compute_per_token_components(
...
-            )
+            is_clipped = (is_lower_clipped & (expanded_advantages < 0)) | (is_upper_clipped & (expanded_advantages > 0))
         else:  # sequence level
             # For sequence level, coef_1 is shape (B, 1), advantages is shape (B,)
-            seq_advantages = advantages
-            is_clipped = ((coef_1.squeeze(-1) < 1 - epsilon_low) & (seq_advantages < 0)) | (
-                (coef_1.squeeze(-1) > 1 + epsilon_high) & (seq_advantages > 0)
-            )
-            is_clipped = is_clipped.unsqueeze(1).expand_as(attention_mask)
+            is_clipped = (is_lower_clipped & (advantages < 0)) | (is_upper_clipped & (advantages > 0))
```
Comment (Tcc0403) [inline test/chunked_loss/test_grpo_loss.py]: Same for test ref model

Comment (Tcc0403) [reply test/chunked_loss/test_grpo_loss.py]:
```suggestion
            is_clipped = (is_lower_clipped & (expanded_advantages < 0)) | (is_upper_clipped & (expanded_advantages > 0))
```
Simply use `expanded_advantages` should be enough

**Comment by yukiu00** [comment]
@Tcc0403 
> Do you mind if I edit your PR directly?

Thank you for catching this . I'm sorry for the oversight 🙏  No, please feel free to edit my PR directly. I appreciate your help!

**Review by Tcc0403 — APPROVED** [review]
Thank you!

**Comment by Tcc0403** [comment]
> Thank you for catching this . I'm sorry for the oversight 🙏 

It's all good. It's your fix help me catch the edge case where `chunk_size != 1`!

**Comment by yukiu00** [comment]
@Tcc0403 

Thank you for merging this! Since I'm planning to use this feature in my project, could you let me know when you expect the next PyPI release to be available?

**Comment by Tcc0403** [comment]
I would suggest use the nightly build if you are in a rush since next release would focus on having full transformers v5 support, but we will consider it!
